### PR TITLE
Don't double up runtime checks

### LINKED ISSUE #8678: EXPORT_ES6 + USE_PTHREADS not compatible with ASSERTIONS
Due to the changes introduced in #8306 EXPORT_ES6 now works with USE_PTHREADS. However I couldn't get the assertions for checking if the runtime is initialised to work in a worker.
There is an assertion added here:
https://github.com/emscripten-core/emscripten/blob/1e5317c764fc94f736b7b38dd8a062f63d8e3308/emscripten.py#L1705
But without that it would throw with `TypeError: Cannot assign to read only property '${overwritten_property_name}' of object '[object Object]'`

As far as I can tell it's only that block of assertion that have trouble.

See this comment chain for more discussion: https://github.com/emscripten-core/emscripten/pull/8306#discussion_r272006611

### PR DESCRIPTION
**Pull request by VirtualTim**
`runtime_assertions = RUNTIME_ASSERTIONS` already adds runtime checks to each function, doing it twice is redundant.

So currently when compiling with `ASSERTIONS=1|2` each function is wrapped in a call that asserts the runtime is initialised. However each function is actually wrapped twice, creating redundant code and breaking strict mode.

For example the Emscripten generated function `__get_daylight` becomes:
```
var __get_daylight = Module["__get_daylight"] = function() {
  assert(runtimeInitialized, 'you need to wait for the runtime to be ready (e.g. wait for main() to be called)');
  assert(!runtimeExited, 'the runtime was exited (use NO_EXIT_RUNTIME to keep it alive after main() exits)');
  return Module["asm"]["__get_daylight"].apply(null, arguments)
};
[...]
var real___get_daylight = asm["__get_daylight"];
asm["__get_daylight"] = function() {
  assert(runtimeInitialized, 'you need to wait for the runtime to be ready (e.g. wait for main() to be called)');
  assert(!runtimeExited, 'the runtime was exited (use NO_EXIT_RUNTIME to keep it alive after main() exits)');
  return real___get_daylight.apply(null, arguments);
};
```
Obviously this is redundant, so this PR removed the second one, so now the generated code is just:
```
var __get_daylight = Module["__get_daylight"] = function() {
  assert(runtimeInitialized, 'you need to wait for the runtime to be ready (e.g. wait for main() to be called)');
  assert(!runtimeExited, 'the runtime was exited (use NO_EXIT_RUNTIME to keep it alive after main() exits)');
  return Module["asm"]["__get_daylight"].apply(null, arguments)
};
```

Fixes #8678

### CODE REVIEW

**Comment by VirtualTim** [comment]
For reference with `ASSERTIONS=0` the call becomes:
```
var __get_daylight= Module["__get_daylight"] = function() {
  return Module["asm"]["__get_daylight"].apply(null, arguments)
};
```

**Comment by VirtualTim** [comment]
I'm not sure why these are failing. Locally they pass for me.
This code is still emitted:
```
var _note = Module["_note"] = function() {
  assert(runtimeInitialized, 'you need to wait for the runtime to be ready (e.g. wait for main() to be called)');
  assert(!runtimeExited, 'the runtime was exited (use NO_EXIT_RUNTIME to keep it alive after main() exits)');
  return Module["asm"]["_note"].apply(null, arguments)
};
```

**Comment by VirtualTim** [comment]
There was a race condition in the test.
`assert(Module.noted);` was called with a 1s timeout, but there was no guarantee that it is always called after `myJSCallback()`.

File: tests/test_browser.py [hunk 1]
```diff
@@ -2331,7 +2331,7 @@ def test_runtime_misuse(self):
         xhr.open('GET', 'http://localhost:%s/report_result?' + HEAP32[Module.noted>>2]);
         xhr.send();
         setTimeout(function() { window.close() }, 1000);
-      }, 1000);
+      }, 2000);
```
Comment (sbc100) [inline tests/test_browser.py]: Can you make this a separate change? 

Makes me sad to see sleep values like this in test code.

File: emscripten.py [hunk 2]
```diff
@@ -1691,27 +1691,9 @@ def create_receiving(function_table_data, function_tables_defs, exported_impleme
...
-};
-''' % {'name': name, 'runtime_assertions': runtime_assertions})
-    receiving = '\n'.join(wrappers)
 
   shared.Settings.MODULE_EXPORTS = module_exports = exported_implemented_functions + function_tables(function_table_data)
 
-  # Currently USE PTHREADS + EXPORT ES6 doesn't work with the previously generated assertions.
-  # ES6 modules disallow re-assigning read-only properties.
-  if shared.Settings.USE_PTHREADS and shared.Settings.EXPORT_ES6 and shared.Settings.ASSERTIONS and receiving:
-    receiving = "assert(!ENVIRONMENT_IS_PTHREAD, 'Error: USE_PTHREADS and EXPORT_ES6 requires ASSERTIONS=0');\n" + receiving
```
Comment (sbc100) [inline emscripten.py]: Is this is an unrelated change?

**Comment by sbc100** [comment]
I agree this is redundant.  I've tried to clean this up before but haven't got around to it.   Since it doesn't effect release builds it not much of a problem but I agree it should be cleanup up.

One thing I noticed is that if you call `Module["asm"]["__get_daylight"]` directly in your code now, won't it skip the assertions?   Since you no long wrapping `Module["asm"]["__get_daylight"]` only `Module["__get_daylight"]`?  I don't know how much of problem that is?  Or if its even legal to call the inner `asm` function?

File: tests/test_browser.py (hunk 1 above)
Comment (VirtualTim) [reply tests/test_browser.py]: Yeah me too. I tried to fix it a different way, but couldn't get everything working, and didn't want to pollute this PR with unrelated stuff. I opened #8939, but I'm not really sure how to fix it.

I was hoping that increasing the time would make it pass, until a proper fix is implemented, but the tests still failed...

File: emscripten.py (hunk 2 above)
Comment (VirtualTim) [reply emscripten.py]: Sort of. I added this assertion originally because I couldn't get `USE_PTHREADS` + `EXPORT_ES6` working with `ASSERTIONS`. This PR fixes the underlying issue, so this isn't needed any more.

**Comment by VirtualTim** [comment]
> One thing I noticed is that if you call `Module["asm"]["__get_daylight"]` directly in your code now, won't it skip the assertions?

Ah I didn't think about that. Yeah I don't know if that's something people actually do. It seems pretty non-standard, but i could be wrong.

File: tests/test_browser.py (hunk 1 above)
Comment (kripken) [reply tests/test_browser.py]: Unless I am missing something, nothing else in the test depends on timing? So the value here could be 1, 1000, 2000, and the result should be the same. If so that makes me suspect it's failing for another reason.

Comment (VirtualTim) [reply tests/test_browser.py]: As far as I can see there's no guarantee that `myJSCallback()` will finish executing before this function is called. So the issue I was seeing was that `assert(Module.noted);` can be hit before it is set (or set to its final value). That's why it was was passing locally for me, but failing on the build system.
Also if `EXIT_RUNTIME=1` is specified, like it is here https://github.com/emscripten-core/emscripten/blob/30f701b18e0e377cca96c0fff36988e453b25486/tests/test_browser.py#L2353 then `myJSCallback()` could actually be called after the runtime has exited.

I think the best thing to do is fix #8939, merge the changes into this PR, then see if the tests still pass.

Comment (kripken) [reply tests/test_browser.py]: I have a possible fix for that in #8964

**Comment by VirtualTim** [comment]
Figured out what the issue was. In `WASM=1` runtime checks are added in via `create_receiving_wasm`, and  `create_receiving`. I don't know if `create_receiving` should be entered at all if compiling with `WASM=0`, but I didn't want to change those function for fear of breaking stuff.

The two failures are gl related, so this should be good to merge, if everyone is happy with the change.

File: emscripten.py [hunk 3]
```diff
@@ -1691,27 +1691,23 @@ def create_receiving(function_table_data, function_tables_defs, exported_impleme
     runtime_assertions = ''
   else:
     runtime_assertions = RUNTIME_ASSERTIONS
-    # assert on the runtime being in a valid state when calling into compiled code. The only exceptions are
-    # some support code
-    receiving_functions = [f for f in exported_implemented_functions if f not in ('_memcpy', '_memset', '_emscripten_replace_memory', '__start_module')]
+    # assert on the runtime being in a valid state when calling into compiled code. The only exceptions are some support code.
+    # WASM=1 already inserts runtime assertions, so no need to do it again here (see create_receiving_wasm)
+    if not shared.Settings.WASM:
```
Comment (sbc100) [inline emscripten.py]: My understanding is that `create_receiving_wasm` is for the wasm backend and `create_receiving` is for fastcomp.   So when running with WASM=1 and fastcomp we still want this code don't we?

File: tests/test_browser.py [hunk 4]
```diff
@@ -2323,19 +2323,19 @@ def test_runtime_misuse(self):
...
     post_hook = r'''
       function myJSCallback() {
+        // Run on the next event loop, as code may run in a postRun right after main().
+        setTimeout(function() {
+          var xhr = new XMLHttpRequest();
+          assert(Module.noted);
+          xhr.open('GET', 'http://localhost:%s/report_result?' + HEAP32[Module.noted>>2]);
+          xhr.send();
+          setTimeout(function() { window.close() }, 1000);
+        }, 0);
```
[the remark this answers is not in the document]
Comment (kripken) [reply tests/test_browser.py]: Merging in incoming should get rid of that, as I landed the PR that provides it.

[the remark this answers is not in the document]
Comment (VirtualTim) [reply tests/test_browser.py]: Yeah that was added in because I merged in this PR #8964 before it was merged into incomming,.

File: emscripten.py (hunk 3 above)
Comment (VirtualTim) [reply emscripten.py]: I'm not sure. That's why I didn't want to restructure the whole thing. What I know is that in `WASM=0` runtime assertions are added once, and in `WASM=1` they are added twice.
I assumed that `create_receiving_wasm` was responsible for that, but I could be wrong.

**Comment by VirtualTim** [comment]
`test_glgears` was the only failure. That test seems to be pretty flaky lately...

File: emscripten.py (hunk 3 above)
Comment (kripken) [reply emscripten.py]: There should be tests for it, but it's possible they are disabled on the wasm backend atm - so we should confirm this before landing.

**Comment by VirtualTim** [comment]
@kripken Any reservations about merging this?

**Comment by kripken** [comment]
@VirtualTim I think it's fine to land now, since I recently triaged all the wasm backend disabled tests, and didn't see any related to these runtime assertions, so we should have full coverage. I also did a quick test and I don't see assertions missing from a hello world program compiled with pthreads or not (diffing against before this PR).

File: emscripten.py (hunk 3 above)
Comment (sbc100) [reply emscripten.py]: From the code seems clear that either `create_receiving_wasm` *or* `create_receiving` is called, never both.  So at the very least this comment it wrong.  Probably worth investigating..
